File: KMIT/Graphs/Problem6.java
/*
Problem 6:

Implement a function to find the number of ways to reach the bottom-right corner of a m x n grid from the top-left
corner, where each cell contains a non-negative integer representing the cost to traverse it. You can only move
down or right at each step.


Example:
Input:
grid = [[1,3,1],
[1,5,1],
[4,2,1]]
Output:
6

*/


import java.util.*;
public class Problem6 { 
    private static int dfs(int[][] grid, int i, int j, int m, int n, int[][] graph) {
        if (i == m - 1 && j == n - 1) {
            return 1;
        }
        if (graph[i][j] != -1) {
            return graph[i][j];
        }
        int ways = 0;
        if (i + 1 < m) {
            ways += dfs(grid, i + 1, j, m, n, graph);
        }
        if (j + 1 < n) {
            ways += dfs(grid, i, j + 1, m, n, graph);
        }
        graph[i][j] = ways;
        return ways;
    }  
    public static int numWays(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int[][] graph = new int[m][n];
        for (int i = 0; i < m; i++) {
            Arrays.fill(graph[i], -1);
        }
        return dfs(grid, 0, 0, m, n, graph);
    }  
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int m = sc.nextInt();
        int n = sc.nextInt();
        int[][] grid = new int[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                grid[i][j] = sc.nextInt();
            }
        }
        System.out.println(numWays(grid));
        sc.close();
    }
}

File: student_Banasthali_Nivedita/graphs/problem6.cpp
/*
6. Implement a function to find the number of ways to reach the bottom-right corner of a m x n grid from the top-left
corner, where each cell contains a non-negative integer representing the cost to traverse it. You can only move
down or right at each step.

*/

#include <iostream>
#include <vector>

using namespace std;

int uniquePaths(vector<vector<int>>& grid) {
    int m = grid.size();
    int n = grid[0].size();

    vector<vector<int>> dp(m, vector<int>(n, 0));

    dp[0][0] = 1;

    for (int j = 1; j < n; j++) {
        dp[0][j] = dp[0][j - 1] + grid[0][j];
    }

    
    for (int i = 1; i < m; i++) {
        dp[i][0] = dp[i - 1][0] + grid[i][0];
    }

   
    for (int i = 1; i < m; i++) {
        for (int j = 1; j < n; j++) {
            dp[i][j] = min(dp[i - 1][j], dp[i][j - 1]) + grid[i][j];
        }
    }

    return dp[m - 1][n - 1];
}

int main() {
    vector<vector<int>> grid = {
        {1, 3, 1},
        {1, 5, 1},
        {4, 2, 1}
    };

    int numWays = uniquePaths(grid);
    cout << "Number of ways: " << numWays << endl;

    return 0;
}

